codalab/server/rest_server.py

```diff
@@ -41,6 +43,77 @@
 ]
 
 
+class RequestCounter(object):
+    def __init__(self, window_seconds):
+        self._lock = threading.Lock()
+        self._counts = {}
+        self._window_seconds = window_seconds
+
+    def count(self, user_id, limit):
+        """
+        Atomically counts requests associated with the given user key.
+        Returns the remaining requests and the remaining seconds in the window
+        since the user's first request.
+        """
+        with self._lock:
+            now = time.time()
+            count, first_time = self._counts.get(user_id, (0, now))
+            time_remaining = int(self._window_seconds - (now - first_time))
+
+            # Check if reset time is reached and reset accordingly
+            if time_remaining > 0:
+                count += 1
+            else:
+                count = 1
+                first_time = now
+                time_remaining = self._window_seconds
+
+            # Update count and prevent integer overflow
+            self._counts[user_id] = (min(count, limit), first_time)
+
+            # Returning remaining count and (truncated) remaining time
+            return limit - count, time_remaining
+
+
+class RateLimitPlugin(object):
+    """
+    Limit the request rate.
+    Request counts are associated with user for authenticated requests,
+    and with IP address for unauthenticated requests. Different limits can be
+    configured for these two cases.
+    Sets the appropriate headers following the GitHub API's conventions.
+    """
+    api = 2
+
+    def __init__(self, window_seconds=60*60, auth_limit=5000, unauth_limit=60):
```

Comment:
A worker can easily hit the 5000 per hour limit. I don't support this change at all. I'd just not do this change and worry about too many requests when we start seeing too many requests.